i18n - extraction tooling

petebacondarwin opened the pull request:
This PR implements a new message extraction tool, which will eventually be integrated into the CLI.

Sits on top of #37114, which should land first.

FW-2073

Closes #37202

ocombe commented on packages/localize/src/tools/src/extract/translation_files/json_translation_serializer.ts:
```diff
@@ -0,0 +1,23 @@
+  renderFile(messages: ParsedMessage[]): string {
+    let fileObj: SimpleJsonTranslationFile = {locale: 'en', translations: {}};
+    messages.forEach(
+        message => { fileObj.translations[message.messageId] = message.messageString; });
```
should you also add the description / meaning inside of the message so that it can be used by translators?
something like: `:meaning:desc:Content of the message` ?

petebacondarwin replied:
Really this JSON based serialization is a super simple example of how to serialize. (I did name it `SimpleJsonTranslationFile`!)

I didn't really expect it to be used in practice. The other formats do add in this metadata about the messages.

I think we should expose a way for 3rd parties to provide their own parser/serializer pairs for other formats rather than try to make this particular one work better.

ocombe replied:
yes, that would be ideal :)

thetric commented:
FYI #32881 was merged 5 days ago in https://github.com/angular/angular/commit/2cdb3a079dff19f3ca93c242c2aaa78cf724a3a2

ocombe commented on packages/localize/src/tools/src/extract/main.ts:
```diff
@@ -0,0 +1,87 @@
+  });
+}
+
+type TranslationFormat = 'json' | 'xmb' | 'xliff1' | 'xliff2';
```
you should also support `xlf` and `xlf2` (and change the switch in `getTranslationSerializer` to support those values)

petebacondarwin replied:
Actually I am redesigning it so that each serializer can determine whether it can handle a format or not, rather than deciding at the top level. This will make it easier in the future to be able to plugin new serializers.

petebacondarwin committed:
feat(localize): implement message extraction tool

This tool, which can be run from the node_modules bin folder, can parse
the source files in your compiled app and generate a translation file
formatted with the configured syntax.

For example:

```
./node_modules/.bin/localize-extract -s 'dist/**/*.js' -f xliff1 -o dist/messages.en.xlf
```

petebacondarwin committed:
test(localize): add compile time extraction to integration test

The integration test for i18n now makes use of the new extraction tooling
from the `@angular/localize` package rather than the old ViewEngine extractor.

BojanKogoj commented:
What's the status of this? We can't migrate to Angular 9 until this is released.

ocombe commented:
Extraction from templates already work with `ng xi18n` from the Angular CLI.

BojanKogoj commented:
Aware of that. But we'd like to replace i18n-polyfill as well, which means we need extraction from components as well.

ocombe commented:
Oh, then you can use `@locl/cli`  with the `extract` command in the mean time
https://github.com/loclapp/locl/tree/master/libs/cli

luchsamapparat commented:
but be aware, that Locl is only free for open source projects!

> You can use them for free in open source projects, but if you want to use them in commercial products with close source you are required to buy a proprietary license (coming soon).
https://github.com/loclapp/locl#license

ricardosaracino commented:
putting this here if someone is stuck

> If someone bumps into this issue, this might be helpful.

> 
> What I did is I've created a TranslationComponent, in which I'll declare all the code translations, ex.:
> 
> <span i18n>Some translation</span>
> 
> Then when in code I write:
> 
> $localize`Some translation`
> everything works "out of the box" because ID is the same.

https://github.com/angular/angular-cli/issues/16375#issuecomment-570351093

JanneHarju commented:
Please hurry! This is critical part of any project. So far this seems to be big step forward.

ricardosaracino commented:
> Please hurry! This is critical part of any project. So far this seems to be big step forward.

I put a decent solution on stackoverflow 

https://stackoverflow.com/questions/60167454/angular-9-localize-not-working-with-configuration-de/60724920#60724920

BojanKogoj commented:
@ricardosaracino That is not exactly a good solution. It works, but this means you need to manage translations in templates as well. You will end up with unused and missing translations.

wincyb commented:
Hi - awaiting for this extract option, please let know with which release(minor) we can expect this?

wincyb commented:
Hi - Please let know in which minor verison we can expect the extract to work for ts files also?

petebacondarwin commented:
Expect it in v10.0.0, out later this year.

wincyb commented:
Thanks for your quick reply @petebacondarwin, when this gets released with Angular 10, will it support extracting to .json file format also like how locl does? please let know, because we are already seeing huge size with xlf files(xml) in prod bundles.

petebacondarwin commented:
@wincyb  - I am not sure what that format is... could you elaborate?

wincyb commented:
> @wincyb - I am not sure what that format is... could you elaborate?

My apologies for the late reply, currently Angular supports xmb|xlf|xlif|xliff|xlf2|xliff2 formats of translation files, so along with Angular 10 will it support json also like how the locl does? since it will reduce build size, please let know

petebacondarwin commented:
Do you have a specification for the layout of the JSON?

ricardosaracino commented:
> My apologies for the late reply, currently Angular supports xmb|xlf|xlif|xliff|xlf2|xliff2 formats of translation files, so along with Angular 10 will it support json also like how the locl does? since it will reduce build size, please let know

Angular i18n makes produces builds per locale... i fail to see how build size is affected by the source file format used. If you dont want multiple baseHrefs ... then i dont think your concern is JSON but the fact you are using Angular i18n

![image](https://user-images.githubusercontent.com/759253/79452707-917ec080-7fb6-11ea-85d4-73751a1b04c1.png)

saidsoualhi commented:
Is there any news about the extracting tooling ? Really we need this asap :(

JohannesHoppe commented:
> @saidsoualhi: Is there any news about the extracting tooling ? Really we need this asap :(

yes, see my comment at https://github.com/angular/angular-cli/issues/17319#issuecomment-626936585

vekunz commented:
As @petebacondarwin mentioned above, you can expect it in Angular 10, not earlier.

petebacondarwin committed:
feat(compiler-cli): add `SourceFile.getOriginalLocation()` to sourcemaps package

This method will allow us to find the original location given a
generated location, which is useful in fine grained work with
source-mapping. E.g. in `$localize` tooling.

TKul6 commented:
Hi, is there any news about this tool?
Is there any chance this tool is shipped with angular 9.1.0?

Thanks

petebacondarwin commented:
Hoping for v10.1 - sorry about the wait.

TKul6 commented:
Hi @petebacondarwin ,
I saw that the id of trans-unit is not unique based on the input value.
For example, lets say that in my component.html I have:
`<div i18n>Chuck Norris can slam a revolving door.</div>`
and in the component.ts I have:
```chuckNoris = $localize`Chuck Norris can slam a revolving door.`;```

when I'll use `xi18n` I'll get for the html template something like: `2d61f30137aa2065553764bd3a7649718a163174`.
However when I will serve the application I'll get a different id (I thing it is a different format), something like: `1795824228003639006`.

It's a bit strange to me, because if I'm not mistaken Angular replace all i18n with $localize calls so it's suppose to be exactly the same.

Is your new tool going to solve this issue?
And do you know why the ids are different (just curious).

Thanks a lot.

p.s. thanks a lot for your hard work, really expect to see the final result.

petebacondarwin commented:
@TKul6 - you have found out my little secret. Well it is not so little, or secret.
The number you see associated with the `$localize` tag is the new message ID format that we are hoping to migrate everyone to in the future.

This ID can be computed purely from the `$localize` tagged template string, so is stable for using in the translation merging/inlining process which runs after compilation and only these tagged strings are left.  The current (legacy) message IDs, of which there are at least two flavours, are dependent upon information (such as whitespace) in the original template that does not get passed through to the `$localize` tagged strings. So we cannot compute them at translation merging/inlining time.

For the time being the Angular compiler annotates `$localize` tagged strings that are generated from `i18n` tagged templates with the legacy message ids so that they are available at translation merging/inlining time. You might have noticed these prepended to the "pre-translated" messages generated in the Ivy templates. For example:

```ts
$localize`:some\:description␟f67d33573c0e555c44bacc91043bef723bdd184b␟2354794512298129537: Hello ${"\ufffd0\ufffd"}:INTERPOLATION:! `
```

The two legacy ids are `f67d33573c0e555c44bacc91043bef723bdd184b` and `2354794512298129537`.

Once this PR lands we can start generating translation files with the new message ID format and then we can turn off legacy messages by setting `"enableI18nLegacyMessageIdFormat": false` in your tsconfig.json file. But this will be a big breaking change for translation files that were already generated in the past by the ViewEngine extraction.

You can read more about this in my design document here: https://hackmd.io/EQF4_-atSXK4XWg8eAha2g

BojanKogoj commented:
> But this will be a big breaking change for translation files that were already generated in the past by the ViewEngine extraction.

@petebacondarwin translations with custom id should continue to work?

We currently use the default Angular i18n (localize now) for template and [i18n-polyfill](https://github.com/ngx-translate/i18n-polyfill) for code translation, do you expect any issues?

Is interpolation changing as well (it looks so in the document)? Current XML-like syntax is giving us a headache because [Poeditor](poeditor.com) removes it for some reason.

petebacondarwin commented:
@BojanKogoj - custom ids are unaffected by this work. Any change we make to the message ID format will include discussion with the community and we will do everything possible to make the transition easier. In any case there will be at least a full major version where this is hidden behind a flag and another full major version where there will be the option to go back to the legacy mode.

> Is interpolation changing as well 
I am not quite sure what you mean by this. Can you elaborate?

TKul6 commented:
Hi @petebacondarwin 

> You can read more about this in my design document here: https://hackmd.io/EQF4_-atSXK4XWg8eAha2g

First of all I want to thank you for the documentation you provided, it shed some light on what happens behind the hood in the localize process.

So if I understand correctly, If I'll take the following actions, I won't experience the breaking changes regarding the message id:

1. When I extract the messages, use **xliff2** instead of the default.
2. If I have `$localize` call in my ts file like: ```$localize`Welcome ${username}:name:` ``` I can get the xliff id matching the string `Welcome {$name}` (which basically what the cli warns me about when I try to serve the application without the appropriate xlf file) and add this entry to the xlf file.

Did I get it right?
Thank you very much, and keep up the good work!

petebacondarwin commented:
I'm afraid that even using XLIFF 2 will not avoid the breaking change. In the document I wrote:

> Both XLIFF 2 and XMB/XTB use the same digest function. The new common digest function should use the same hashing function as these but compute the canonical message string in a way that is resilient to whitespace changes (if appropriate) and can be computed from the information provided to $localize alone.

Note that even though they are using the same "hashing" function, the input this function will be slightly different because of the change to how whitespace is handled. This is why, currently, we add both hashes to the generated `$localize` tagged strings.

That being said the second of your points, regarding ids of `$localize` tagged strings in your own TS files is correct. These do not have to support the legacy message ids since there was no such thing pre-$localize. So these IDs will stay the same before and after the migration from legacy message ids.

BojanKogoj commented:
>I am not quite sure what you mean by this. Can you elaborate?

@petebacondarwin  The following

```
<source>Hello <x id="INTERPOLATION" equiv-text="{{ user?.name }}"/>, welcome back!</source>
```

This will only display `Hello , welcome back!` in some translation tools, such as poeditor (it completely removes it)

petebacondarwin commented:
@BojanKogoj - I am not sure what POEditor is doing but this is accurate XML for describing placeholders, which is what an INTERPOLATION is. If it is not working as expected in the editing tool then I would contact them for support.

petebacondarwin committed:
refactor(localize): handle location endings correctly

Previously source locations required an ending position but this was not
being computed effectively. Now ending position is optional and it is
computed from an `endPath` passed to `getLocation()`.

TKul6 commented:
Hi @petebacondarwin , I tried to extract all the `$localize` calls and create trans unit from them:
```<trans-unit id="4114913878503799192">
        <source>No results were found</source>
        <target>PREFIX No results were found</target>
      </trans-unit>
```

It works fine except 1 issue, if I have place holders there like: `"The destination already has a folder named {$folderName}."` it failed to populate the values in the string.

The trans unit is: 
```<trans-unit id="6820689397495822812">
        <source>The destination already has a folder named &lt;x id="INTERPOLATION" equiv-text="folderName" /&gt;.</source>
        <target>PREFIX The destination already has a folder named &lt;x id="INTERPOLATION" equiv-text="folderName" /&gt;.</target>
        <note>modified source.</note>
      </trans-unit>
```

Do you have any idea why it fails to parse the interpolated values (please don't say this behavior is yet to be implemented).

Thanks

petebacondarwin commented:
@TKul6 - the interpolation elements are part of the XLIFF syntax. Did you mean to escape them?

It should look like:

```
        <source>The destination already has a folder named <x id="INTERPOLATION" equiv-text="folderName" />.</source>
        <target>PREFIX The destination already has a folder named <x id="INTERPOLATION" equiv-text="folderName" />.</target>
        <note>modified source.</note>
      </trans-unit>
```

petebacondarwin committed:
fix(compiler-cli): ensure source-maps can handle webpack:// protocol

Webpack and other build tools sometimes inline the contents of the
source files in their generated source-maps, and at the same time
change the paths to be prefixed with a protocol, such as `webpack://`.

This can confuse tools that need to read these paths, so now it is
possible to provide a mapping to where these files originated.

TKul6 commented:
Hi @petebacondarwin,
I made it work! (When I use the xliff 2 format I'm able to take the error messages the cli emits and create the trans-units based on that data).

Thank you (looking forward for the extraction tool!)

rockument69 commented:
@petebacondarwin : I have added a "helpful" callout into my PR [36924](https://github.com/angular/angular/pull/32912) for refactoring the existing version of i18n.md:

> <div class=“alert is-helpful”>
> 
> The ng xi18n extraction mechanism uses the ViewEngine compiler, which is not able to cope with some code syntax that the Ivy compiler can handle. A fix for this issue is currently under development. To follow or contribute to this fix see [PR 32912](https://github.com/angular/angular/pull/32912).
> 
> </div>

When this PR is closed, please delete the above callout. Thanks!

TKul6 commented:
Hi @petebacondarwin ,
Small thing that bother me with the current extraction tool: **There is no easy way to merge the existing xlf files with the new results from `xi18n`**.

Is the new extraction tool will support that?

petebacondarwin commented:
The initial implementation is aiming for feature parity with our current tooling (plus `$localize` in app code support).
We can then look at translation file merging, although this might be a bit of stretch for our team's resources, and it might be that this could better be achieved via a 3rd party tool, if someone was keen to build such a merge tool.

laurantines commented:
Hello everyone!

I don't know if this is the right place to explain what happens to me when trying to extract all the text strings that I have in my project with Angular 9. For this I use the command npm run extract-i18n
and extract in XLIFF the strings of the HTML templates (I know that the $localize of TypeScript files must be done manually.) But I get an error that Tagged template expressions are not supported in metadata in 'routes' and it is related to adding $localizebreadcumb name and $localizetitle name. In other words, when I try to translate my menu or the breadcrumbs and I run so that it extracts the other strings from my HTML, I skip that message and it doesn't generate anything for me. Before translating the menus and breadcrumbs I did it right. Any ideas? Seems related to @angular/localize library.

Best regards!

petebacondarwin commented:
@laurantines - the problem you have is that our current extraction tooling relies upon the ViewEngine compiler, which is not able to cope with these tagged strings. There should be no problem for you once this PR lands and is implemented into the CLI.

petebacondarwin committed:
refactor(compiler): add source-map spans to localized strings

Previously localized strings were not mapped to their original
source location, so it was not possible to back-trace them
in tools like the i18n message extractor.

petebacondarwin committed:
fix(compiler-cli): only read source-map comment from last line

Source-maps can be linked to from a source-file by a comment at
the end of the file.

Previously the `SourceFileLoader` would read
the first comment that matched `//# sourceMappingURL=` but
this is not valid since some bundlers may include embedded
source-files that contain such a comment.

Now we only look for this comment in the last non-empty line
in the file.

josephperrott commented on integration/ivy-i18n/package.json:
```diff
@@ -45,8 +46,8 @@
   "devDependencies": {
-    "@angular-devkit/build-angular": "file:../../node_modules/@angular-devkit/build-angular",
-    "@angular/cli": "file:../../node_modules/@angular/cli",
+    "@angular-devkit/build-angular": "0.1000.0-rc.3",
```
Is there a reason we prefer to only move this integration test forwards on CLI as opposed to updating the base `package.json`'s version?

petebacondarwin replied:
@josephperrott  - Oh yes! I meant to comment about this... good spot.

I was having a terrible time with a hoisting problem, which only seems to happen when the dependency is coming from the top level node_modules folder.
If I remember correctly it was that the version of `mkdirp` that was being placed as a dependency of `build-angular` was the wrong major version. And this package had a major change going from being callback-based to promise-based.
I wondered if we might be able to avoid this by doing an `npm pack` on the top level installed packages and then installing those in the integration tests...
Alternatively we could have a script that updated all the `package.json` files for the integration tests whenever the top level `package.json` changed...
Do you have any other ideas? Or a way to fix this?

josephperrott replied:
Gotcha, so the issue comes because we are just taking the installed files from node modules, not saying to use the values from the base `package.json`.

Long term we could look at adding support for our `npm_integration_test` to support the same logic we have in `//dev-infra`, where you can specify a version as `<from-root>`.  This substitutes in the version number from the root `package.json`, so that all of the packages dependencies are properly met.

In the short term, we can land this as is, rather than requiring that the entire repo goes through an update and migration to handle the version change.

AndrewKushnir commented on packages/localize/src/tools/src/extract/extraction.ts:
```diff
@@ -0,0 +1,94 @@
+      const originalEnd = location.end !== undefined ?
+          sourceFile.getOriginalLocation(location.end.line, location.end.column) :
+          null;
+      if (originalEnd !== null && originalEnd.file === originalStart.file) {
```
I think it'd be great to add a quick comment here to describe in which cases `originalEnd.file` may not be the same as `originalStart.file`.

petebacondarwin replied:
I'll add a comment, although I confess I don't really know in what circumstances this might occur... perhaps if a build tool went berserk and corrupted the source-map??

This is mainly here for sanity, since it would make no sense to have line number from different files, when there is only the option of describing a single file in the translation file formats.

AndrewKushnir commented on packages/localize/src/tools/src/extract/main.ts:
```diff
@@ -0,0 +1,149 @@
+            describe: 'The lowest severity logging message that should be output.',
+            choices: ['debug', 'info', 'warn', 'error'],
+          })
+          .option('use-source-maps', {
```
The `use-source-maps` and `use-legacy-ids` param names format is different from `basePath` and `outputPath`, can we use the same format?

petebacondarwin replied:
Good call!

AndrewKushnir commented on packages/localize/src/tools/src/extract/main.ts:
```diff
@@ -0,0 +1,149 @@
+
+  const basePath = options['b'];
+  const sourceFilePaths = glob.sync(options['s'], {cwd: basePath, nodir: true});
+  const logLevel = options['l'] as keyof typeof LogLevel | undefined;
```
nit: I'd propose to put `()` around `keyof typeof LogLevel` for readability.

AndrewKushnir reviewed (commented):
Hi @petebacondarwin, thanks for implementing this feature 👍 

I went though ~60% and left a few comments: mostly minor notes/comments. I will complete CR tomorrow and let you know.

Thank you.

JoostK commented on packages/compiler-cli/src/ngtsc/sourcemaps/src/source_file.ts:
```diff
@@ -87,6 +87,41 @@ export class SourceFile {
+    }
+
+    const locationSegment: SegmentMarker =
+        {line, column, position: this.startOfLinePositions[line] + column, next: undefined};
```
Would it be wise to verify that `line` is within bounds?

petebacondarwin replied:
Yes I think so. Just in case :-)

JoostK commented on packages/compiler-cli/src/ngtsc/sourcemaps/src/source_file_loader.ts:
```diff
@@ -168,6 +170,23 @@ export class SourceFileLoader {
+   */
+  private resolveSourcePath(sourceRoot: AbsoluteFsPath, sourcePath: string): AbsoluteFsPath {
+    const protocolMappedPath = sourcePath.replace(
+        /^([a-z]+):\/\//, (path, protocol) => this.protocolMap[protocol] || path);
```
Should this be case insensitive?

```suggestion
        /^([a-z]+):\/\//i, (path, protocol) => this.protocolMap[protocol] || path);
```

petebacondarwin replied:
You are right!

> Although schemes are case-insensitive, the canonical form is lowercase and documents that specify schemes must do so with lowercase letters.

JoostK commented on packages/compiler-cli/src/ngtsc/sourcemaps/test/source_file_loader_spec.ts:
```diff
@@ -119,7 +119,7 @@ runInEachFileSystem(() => {
            const indexSourceMap = createRawSourceMap({
              file: 'index.js',
-             sources: ['x.js', 'y.js', 'z.js'],
+             sources: ['x.js', 'y.js', 'webpack:///src/z.js'],
```
Can you add a test for when there's no protocol mapping available?

petebacondarwin replied:
Actually this brings up a good point! Currently without any mapping we get this somewhat unwanted situation:

```xml
<context context-type="sourcefile">../../ivy-i18n/webpack:/src/app/app.component.ts</context>
```

I hoped to avoid such as situation... apart from anything else the typings as they stand were nice and clean - that internal paths to sources were always absolute (i.e. started with `/`) but to support this we would need an additional type that allowed "absolute URLs" such as `webpack://...`...

Or perhaps, actually, if no mapping is found we should just strip out the `scheme://` bit and leave just the path... in other words `webpack:///src/z.js` would become `/src/z.js`....

petebacondarwin replied:
That is true but the contents can be (and usually are) provided inline in the source-map. The point of having this path is more informational. I think that `null` throws away potential information for the end-user...

I think, ideally, if we cannot map it then we should retain the  ~~protocol~~scheme but it just makes the typings here really messy as it will touch lots of the system.

JoostK commented on packages/localize/src/tools/src/source_file_utils.ts:
```diff
@@ -347,15 +348,31 @@ export function buildCodeFrameError(path: NodePath, e: BabelParseError): string
+  const location: ɵSourceLocation = {start: getLineAndColumn(startLocation.start), file};
+  const endLocation =
+      endPath && getFileFromPath(endPath) === file && endPath.node.loc || startLocation;
+  if (endLocation) {
```
This check seems redundant at the moment, as `startLocation` is used as fallback for `endLocation`. Now that you made `ɵSourceLocation.end` optional I would expect the fallback `startLocation` to be removed (so this check stays relevant in that case)

JoostK commented on packages/localize/src/tools/test/source_file_utils_spec.ts:
```diff
@@ -5,13 +5,15 @@
 import template from '@babel/template';
 import {Expression, Identifier, TaggedTemplateExpression, ExpressionStatement, FunctionDeclaration, CallExpression, isParenthesizedExpression, numericLiteral, binaryExpression, NumericLiteral} from '@babel/types';
 import {isGlobalIdentifier, isNamedIdentifier, isStringLiteralArray, isArrayOfExpressions, unwrapStringLiteralArray, unwrapMessagePartsFromLocalizeCall, wrapInParensIfNecessary, buildLocalizeReplacement, unwrapSubstitutionsFromLocalizeCall, unwrapMessagePartsFromTemplateLiteral, getLocation} from '../src/source_file_utils';
+import {absoluteFrom} from '@angular/compiler-cli/src/ngtsc/file_system';
```
nit: move to top

JoostK commented on packages/compiler-cli/src/ngtsc/sourcemaps/src/source_file_loader.ts:
```diff
@@ -171,21 +173,34 @@ export class SourceFileLoader {
+           contents[lastRealLineIndex] === '\r') {
+      lastRealLineIndex--;
+    }
+    lastRealLineIndex = contents.lastIndexOf('\n', lastRealLineIndex - 1);
```
I'm confused here. `lastRealLineIndex` is a line number, whereas `lastIndexOf` searches the buffer which needs character offsets. I am missing something or is this incorrect?

petebacondarwin replied:
`lastRealLineIndex` is the character position of a `\n` or `\r` character at the end of the file. In this statement we are searching backwards from that position to find the previous `\n` character position.

So if we had a `contents` that looked like `"line 1\nline 2\n\n\n"` then `lastRealLineIndex` starts out as `13` but then this statement will change it to `6`.

petebacondarwin replied:
I'll rename

JoostK reviewed (commented):
With two commits remaining some initial comments.

- Typo in "ensure source-maps can handle webpack:// protocol"
`This can confuse tool` -> `This can confuse tools`

JoostK commented on packages/localize/src/tools/test/extract/translation_files/xliff1_translation_serializer_spec.ts:
```diff
@@ -0,0 +1,80 @@
+  describe('Xliff1TranslationSerializer', () => {
+    [false, true].forEach(useLegacyIds => {
+      describe(`renderFile() [using ${useLegacyIds ? 'legacy' : 'canonical'} ids]`, () => {
+        it('should convert a set of parsed messages into a XML string', () => {
```
```suggestion
        it('should convert a set of parsed messages into an XML string', () => {
```

JoostK commented on packages/localize/src/tools/test/extract/translation_files/xliff2_translation_serializer_spec.ts:
```diff
@@ -0,0 +1,98 @@
+  describe('Xliff2TranslationSerializer', () => {
+    [false, true].forEach(useLegacyIds => {
+      describe(`renderFile() [using ${useLegacyIds ? 'legacy' : 'canonical'} ids]`, () => {
+        it('should convert a set of parsed messages into a XML string', () => {
```
```suggestion
        it('should convert a set of parsed messages into an XML string', () => {
```

JoostK commented on packages/localize/src/tools/test/extract/translation_files/xmb_translation_serializer_spec.ts:
```diff
@@ -0,0 +1,53 @@
+  describe('XmbTranslationSerializer', () => {
+    [false, true].forEach(useLegacyIds => {
+      describe(`renderFile() [using ${useLegacyIds ? 'legacy' : 'canonical'} ids]`, () => {
+        it('should convert a set of parsed messages into a XML string', () => {
```
```suggestion
        it('should convert a set of parsed messages into an XML string', () => {
```

JoostK commented on packages/localize/src/tools/test/source_file_utils_spec.ts:
```diff
@@ -213,10 +213,10 @@ describe('utils', () => {
         const taggedTemplate = getTaggedTemplate('const x = $localize ``;', {sourceRoot: '/root'});
         const location = getLocation(taggedTemplate);
-        expect(location).toBe(null);
+        expect(location).toBeUndefined();
```
Should this change be in an earlier commit?

JoostK commented on packages/localize/src/tools/src/extract/translation_files/xmb_translation_serializer.ts:
```diff
@@ -0,0 +1,97 @@
+        `<!ELEMENT ex (#PCDATA)>\n` +
+        `]>\n`);
+    xml.startTag('messagebundle');
+    messages.forEach(message => {
```
Same comments as before :-)

JoostK commented on integration/ivy-i18n/package.json:
```diff
@@ -23,10 +23,11 @@
+    "translated:legacy-xlf:update-translations": "node update-xlf-translation-file.js \"../tmp/legacy-locales/messages.legacy.xlf\"",
+    "translated:legacy-xmb:test": "yarn ng build && yarn extract -f xmb -o ../tmp/legacy-locales/messages.legacy.xmb && yarn translated:legacy-xmb:update-translations && yarn ng build --configuration=translated-legacy-xmb && yarn translated:legacy:translate -t \"../tmp/legacy-locales/messages.legacy.xtb\" && yarn translated:legacy:e2e",
+    "translated:legacy-xmb:update-translations": "node update-xmb-translation-file.js \"../tmp/legacy-locales/messages.legacy.xmb\"",
+    "extract": "localize-extract -b . -s dist/main-es2015*.js"
```
I prefer using the long-hand option in scripts so that it's much easier to see what's going without having to know what all options mean.

JoostK commented on integration/ivy-i18n/update-xlf-translation-file.js:
```diff
@@ -0,0 +1,16 @@
+const fs = require('fs');
```
Can you include a comment for this file and the XMB file what is being doing here, and why?

petebacondarwin commented on packages/localize/src/tools/src/extract/extraction.ts:
```diff
@@ -0,0 +1,94 @@
+  private updateLocation(sourceFile: SourceFile, location: ɵSourceLocation): void {
+    const originalStart =
+        sourceFile.getOriginalLocation(location.start.line, location.start.column);
+    if (originalStart !== null) {
```
I have just been rewriting this function. It will now have early returns but also will not mutate the `location`.

petebacondarwin commented on packages/localize/src/tools/src/extract/translation_files/json_translation_serializer.ts:
```diff
@@ -0,0 +1,25 @@
+
+interface SimpleJsonTranslationFile {
+  locale: string;
+  translations: Record<string, string>;
```
I can do better than that - using types:

```
import {ɵMessageId, ɵParsedMessage, ɵSourceMessage} from '@angular/localize';

...

  translations: Record<ɵMessageId, ɵSourceMessage>;
```

petebacondarwin commented on packages/localize/src/tools/src/extract/translation_files/json_translation_serializer.ts:
```diff
@@ -0,0 +1,25 @@
+import {ɵParsedMessage} from '@angular/localize';
+import {TranslationSerializer} from './translation_serializer';
+
+interface SimpleJsonTranslationFile {
```
It is definitely not an official standard. I just made it up last year to hold translations in a simple serialisable object format. It is not really exposed publicly.

By the way there is a "kind of" official JSON format called ARB. See https://github.com/angular/angular/pull/36795

petebacondarwin commented on packages/localize/src/tools/src/extract/translation_files/translation_serializer.ts:
```diff
@@ -0,0 +1,19 @@
+   * Render the contents of a translation file containing the given `messages`.
+   * @param messages The messages to render to the file.
+   */
+  renderFile(messages: ɵParsedMessage[]): string;
```
That's a very reasonable suggestion!

petebacondarwin commented on packages/localize/src/tools/src/extract/translation_files/json_translation_serializer.ts:
```diff
@@ -0,0 +1,25 @@
+  constructor(private sourceLocale: string) {}
+  renderFile(messages: ɵParsedMessage[]): string {
+    let fileObj: SimpleJsonTranslationFile = {locale: this.sourceLocale, translations: {}};
+    messages.forEach(message => {
```
Not sure... I wrote this bit in September 2019!!

petebacondarwin commented on packages/localize/src/tools/src/extract/main.ts:
```diff
@@ -0,0 +1,149 @@
+          .option('f', {
+            alias: 'format',
+            required: true,
+            choices: ['xmb', 'xlf', 'xlif', 'xliff', 'xlf2', 'xliff2'],
```
Oversight. I'll add

petebacondarwin commented on packages/localize/src/tools/src/extract/main.ts:
```diff
@@ -0,0 +1,149 @@
+            alias: 'basePath',
+            required: true,
+            describe:
+                'The base path for paths to the source files and output paths. This should either be absolute or relative to the current working directory.'
```
Changing to:

```
The base path for other paths provided in these options.
This should either be absolute or relative to the current working directory.
```

petebacondarwin commented on packages/localize/src/tools/src/extract/main.ts:
```diff
@@ -0,0 +1,149 @@
+  });
+}
+
+export interface ExtractTranslationsOptions {
```
added

petebacondarwin commented on packages/localize/src/tools/src/extract/translation_files/xliff1_translation_serializer.ts:
```diff
@@ -0,0 +1,99 @@
+    messages.forEach(message => {
+      const id = this.getMessageId(message);
+      if (ids.has(id)) {
+        // Do not render the same message more than one
```
Added context and links

petebacondarwin replied:
Messages with the same `id` are considered identical (whatever their source location, which is purely informational metadata). I don't believe that we can have more than one location per message... Certainly this is not supported in the ViewEngine extractor. If it becomes necessary it could be added later.

JoostK replied:
That sounds fair enough

petebacondarwin commented on packages/localize/src/tools/src/extract/translation_files/xliff1_translation_serializer.ts:
```diff
@@ -0,0 +1,99 @@
+   */
+  private getMessageId(message: ɵParsedMessage): string {
+    return this.useLegacyIds && message.legacyIds !== undefined &&
+        message.legacyIds.find(id => id.length === 40) ||
```
C'est la vie! No other way to tell them apart. Since this is just a legacy hack, which will eventually be removed, it will not be extended with additional legacy message id types. So we are safe to use this.

Updated to be a constant to avoid the magic inline constant.

petebacondarwin commented on packages/localize/src/tools/src/extract/translation_files/xliff2_translation_serializer.ts:
```diff
@@ -0,0 +1,107 @@
+      'srcLang': this.sourceLocale
+    });
+    xml.startTag('file');
+    messages.forEach(message => {
```
sorted!

petebacondarwin commented on packages/localize/src/tools/src/extract/translation_files/xml_file.ts:
```diff
@@ -0,0 +1,113 @@
+    return this;
+  }
+
+  startPreserveWhitespace() {
+    this.preservingWhitespace = true;
+  }
+  stopPreserveWhitespace() {
+    this.preservingWhitespace = false;
+  }
```
indeed

petebacondarwin replied:
removing

JoostK commented on packages/localize/src/tools/src/extract/translation_files/xmb_translation_serializer.ts:
```diff
@@ -40,27 +48,27 @@ export class XmbTranslationSerializer implements TranslationSerializer {
+    for (const message of messages) {
       const id = this.getMessageId(message);
       if (ids.has(id)) {
         // Do not render the same message more than one
```
Missed the one once :-)

<sup>Pun intended, for once</sup>

petebacondarwin commented on packages/localize/src/tools/src/extract/main.ts:
```diff
@@ -27,22 +27,23 @@ if (require.main === module) {
             describe:
-                'An glob pattern indicating what files to search for translations, e.g. `./dist/**/*.js`. This should be relative to the basePath.',
+                'A glob pattern indicating what files to search for translations, e.g. `./dist/**/*.js`.\n' +
+                'This should be relative to the basePath.',
```
Doh!

petebacondarwin commented on packages/localize/src/tools/src/extract/main.ts:
```diff
@@ -55,50 +56,78 @@ if (require.main === module) {
+   */
   basePath: string;
+  /**
+   * A glob pattern indicating what files to search for translations, e.g. `./dist/*.js`. This
+   * should be relative to the basePath.
```
Ah. I copy-pasted the comments. Mea culpa

gkalpak reviewed (approved):
Commit message typo in **refactor(ngcc): move `sourcemaps` into `ngtsc`**:
- `` (such as `@angular/localize`. `` --> `` (such as `@angular/localize`). ``

Reviewed-for: dev-infra, fw-ngcc

AndrewKushnir commented on packages/localize/src/tools/test/extract/integration/main_spec.ts:
```diff
@@ -0,0 +1,226 @@
+    });
+
+    for (const target of ['es2015', 'es5']) {
+      it('should use source-maps to find the original location of translations', () => {
```
nit: may be include `target` into the test description:
```suggestion
      it(`should use source-maps to find the original location of translations (using ${target} bundle)`, () => {
```

AndrewKushnir reviewed (approved):
Great work @petebacondarwin 👍

petebacondarwin commented on packages/compiler-cli/ngcc/index.ts:
```diff
@@ -10,8 +10,7 @@ import {NodeJSFileSystem, setFileSystem} from '../src/ngtsc/file_system';
 
-export {ConsoleLogger} from './src/logging/console_logger';
-export {Logger, LogLevel} from './src/logging/logger';
+export {ConsoleLogger, Logger, LogLevel} from '../src/ngtsc/logging';
```
Actually this is enforced by the formatter. Re-exports must appear below imports

petebacondarwin commented on packages/compiler-cli/src/ngtsc/logging/BUILD.bazel:
```diff
@@ -0,0 +1,14 @@
+package(default_visibility = ["//visibility:public"])
```
Interestingly not in `packages/compiler-cli/BUILD.bazel`, which is the template I was following.

But I see that this was updated in https://github.com/angular/angular/commit/dad5a258b860de83ff6a48e19063151efba24644 to remove a duplicate line - and they removed the one below the load rather than above!!

petebacondarwin replied:
I checked all the packages below `ngtsc` and these ones also had this line at the top of the file: `cycles`, `perf`, `routing`, `scope`, `scope/test`. But most are indeed after the `load()`. And that seems to make sense when coming from the ESM world where imports must be at the top.

So I will change all these.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/logging/BUILD.bazel:
```diff
@@ -0,0 +1,14 @@
+ts_library(
+    name = "logging",
+    srcs = ["index.ts"] + glob([
+        "src/*.ts",
```
OK

petebacondarwin commented on packages/compiler-cli/src/ngtsc/logging/testing/src/mock_logger.ts:
```diff
@@ -6,7 +6,7 @@
  */
 
-import {Logger, LogLevel} from '../../src/logging/logger';
+import {Logger, LogLevel} from '../..';
```
It is intentional. What would be better?

petebacondarwin commented on packages/compiler-cli/src/ngtsc/logging/test/BUILD.bazel:
```diff
@@ -0,0 +1,23 @@
+        "**/*.ts",
+    ]),
+    deps = [
+        "//packages:types",
```
No. Removed.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/logging/testing/BUILD.bazel:
```diff
@@ -0,0 +1,16 @@
+    deps = [
+        "//packages:types",
+        "//packages/compiler-cli/src/ngtsc/logging",
+        "@npm//typescript",
```
No removed

petebacondarwin commented on packages/compiler-cli/ngcc/src/rendering/source_maps.ts:
```diff
@@ -11,8 +11,8 @@ import * as ts from 'typescript';
 import {Logger} from '../../../src/ngtsc/logging';
-import {RawSourceMap} from '../sourcemaps/raw_source_map';
-import {SourceFileLoader} from '../sourcemaps/source_file_loader';
+import {RawSourceMap} from '../../../src/ngtsc/sourcemaps/src/raw_source_map';
+import {SourceFileLoader} from '../../../src/ngtsc/sourcemaps/src/source_file_loader';
```
Good spot - I had to ensure `RawSourceMap` was actually exported from `ngtsc/sourcemaps` too.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/sourcemaps/BUILD.bazel:
```diff
@@ -0,0 +1,19 @@
+        "src/*.ts",
+    ]),
+    deps = [
+        "//packages:types",
```
I think not

petebacondarwin commented on packages/compiler-cli/src/ngtsc/sourcemaps/test/BUILD.bazel:
```diff
@@ -0,0 +1,29 @@
+        "**/*.ts",
+    ]),
+    deps = [
+        "//packages:types",
```
I think not

petebacondarwin commented on packages/compiler-cli/src/ngtsc/sourcemaps/src/source_file_loader.ts:
```diff
@@ -13,6 +13,8 @@ import {Logger} from '../../logging';
 import {RawSourceMap} from './raw_source_map';
 import {SourceFile} from './source_file';
 
+const SCHEME_MATCHER = /^([a-z]+):\/\//i;
```
I didn't realise that schemes can contain non-letters: 

> A non-empty scheme component followed by a colon (:), consisting of a sequence of characters beginning with a letter and followed by any combination of letters, digits, plus (+), period (.), or hyphen (-).

I'll update the regular expression.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/sourcemaps/src/source_file_loader.ts:
```diff
@@ -25,9 +27,16 @@ import {SourceFile} from './source_file';
   private currentPaths: AbsoluteFsPath[] = [];
 
+  /**
+   * Create a new instance of a `SourceFileLoader`.
```
I only wanted to add the JSDOC for the `schemeMap` param so I ended up writing this rubbish.
I just remembered I can add the JSDOC directly to the parameter, so I will ditch this unnecessary boilerplate.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/sourcemaps/src/source_file_loader.ts:
```diff
@@ -181,6 +182,22 @@ export class SourceFileLoader {
+    // webpack).
+    let lastRealLineIndex = contents.length - 1;
+    while (lastRealLineIndex > 0 && contents[lastRealLineIndex] === '\n' ||
+           contents[lastRealLineIndex] === '\r') {
```
Doh!

petebacondarwin commented on packages/compiler-cli/src/ngtsc/sourcemaps/src/source_file_loader.ts:
```diff
@@ -172,20 +182,18 @@ export class SourceFileLoader {
-    const protocolMappedPath = sourcePath.replace(
-        /^([a-z]+):\/\//, (path, protocol) => this.protocolMap[protocol] || path);
-    return this.fs.resolve(sourceRoot, protocolMappedPath);
+  private mapScheme(path: string): string {
```
I'll rename the method to  `replaceSchemeWithPath()` to make it clearer and avoid referencing the implementation so much.

petebacondarwin commented on packages/localize/src/tools/src/extract/main.ts:
```diff
@@ -0,0 +1,149 @@
+import {Xliff1TranslationSerializer} from './translation_files/xliff1_translation_serializer';
+import {Xliff2TranslationSerializer} from './translation_files/xliff2_translation_serializer';
+import {XmbTranslationSerializer} from './translation_files/xmb_translation_serializer';
+import {ɵParsedMessage} from '@angular/localize/private';
```
My convention is that anything inside the `src/tools` folder should be a relative path; and anything above that should be a path-mapped import - and only from `@angular/localize` not ``@angular/localize/private`. So I'll fix up these.

petebacondarwin commented on packages/localize/src/tools/src/extract/source_files/es2015_extract_plugin.ts:
```diff
@@ -0,0 +1,29 @@
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+import {ɵParsedMessage, ɵparseMessage} from '@angular/localize';
```
See https://github.com/angular/angular/pull/32912#discussion_r443118411

petebacondarwin commented on packages/localize/src/tools/src/extract/translation_files/xmb_translation_serializer.ts:
```diff
@@ -0,0 +1,97 @@
+ */
+import {AbsoluteFsPath, relative} from '@angular/compiler-cli/src/ngtsc/file_system';
+import {ɵParsedMessage} from '@angular/localize';
+import {ɵSourceLocation} from '@angular/localize/private';
```
See https://github.com/angular/angular/pull/32912#discussion_r443118411

vekunz commented:
If I write $localize in the data object of a Route, then the xi18n command throws the error "Tagged template expressions are not supported in metadata.". Is this going to work with the new extraction tool?

petebacondarwin commented:
@vekunz if the code builds normally (i.e. outside of the `ng xi18n` command then, yes, this will work in the new setup.

AndrewKushnir commented:
[Presubmit](http://test/OCL:317875890:BASE:317875928:1592928039948:6cf8619a)

AndrewKushnir commented:
FYI, tests in Google's codebase went well (no failures related to these changes). Thank you.

sistla001 commented:
Is this available with 10.0.0? I just updated our project ro 10.0.0. Neither  `ng xi18n` nor `ng i18n-extract ` is extracting localized messages from ts files. Am I missing anything?

BojanKogoj commented:
@sistla001 I don't think so. This was merged after 10.0 release, and they said they [hope it will be in v10.1 anyway](https://github.com/angular/angular/pull/32912#issuecomment-634759439).

offline-pixel commented:
Trying to extract dynamic text 

> ng i18n-extract

Not working for me. However, as text translation can be done on back-end than how it will be benefit a front-end developer? By adding more work??

petebacondarwin commented:
The new extractor is not yet implemented in the CLI, it is only available on the `master` branch of the framework as a standalone binary (`localize-extract`).

sistla001 commented:
> The new extractor is not yet implemented in the CLI, it is only available on the `master` branch of the framework as a standalone binary (`localize-extract`).
Is there a way we can use it to extract translations from typescript files? or should we need to wait for it to be part of the cli?

petebacondarwin commented:
On Wednesday (I think) this week we should release version `10.1.0-next.0`, which will contain the binary you can try. If you are using `yarn` then you can run the command simply by:

```
yarn localize-extract
```

Otherwise you might need to use its actual path:

```
node node_modules/.bin/localize-extract
```

Since this is not integration with the CLI, you will need to pass it some options, and I also recommend building your source with optimisations turned off and source-maps turned on to get the best output.

vekunz commented:
Is there a GitHub issue where we can tract the process of implementing this in the CLI?

petebacondarwin commented:
Not right now. I am working with the CLI team on this...

malichishti commented:
> On Wednesday (I think) this week we should release version `10.1.0-next.0`, which will contain the binary you can try. If you are using `yarn` then you can run the command simply by:
> 
> ```
> yarn localize-extract
> ```
> 
> Otherwise you might need to use its actual path:
> 
> ```
> node node_modules/.bin/localize-extract
> ```
> 
> Since this is not integration with the CLI, you will need to pass it some options, and I also recommend building your source with optimisations turned off and source-maps turned on to get the best output.

@petebacondarwin just to confirm if i understand, are you suggesting that even if it is not part of cli and is perhaps not released i can still use it somehow? If yes, you can please provide a little detailed instructions as i my project is in dire need of this solution.

petebacondarwin commented:
This feature will be in the next "next" release of Angular (i.e. `10.1.0-next.0`) but if you want to access it now you could change your package.json to install the bleeding edge build:

```json
"devDependencies": {
  ...
  "@angular/localize": "https://github.com/angular/localize-builds",
  ...
}
```

Then you should be able to run `node node_modules/.bin/localize-extract`.

malichishti commented:
Thank you @petebacondarwin, i tried it but it gave this:
![image](https://user-images.githubusercontent.com/17155996/86545482-ad20b400-bf72-11ea-85dd-3c7e1cf56823.png)

petebacondarwin commented:
Ah, try it without the `node` prefix...

```
node_modules/.bin/localize-extract
```

Since you are on Windows, there is probably a cmd script there

malichishti commented:
Thank you @petebacondarwin, unfortunately i now get below:
![image](https://user-images.githubusercontent.com/17155996/86693519-16292a00-c04e-11ea-8ea1-1e6d6c3ac88b.png)

petebacondarwin commented:
Ahh, this is because there is now a version mismatch... `@angular/localize@next` depends upon `@angular/compiler-cli@next`. So you need to update the compiler-cli too. Sorry, I forgot about that.

sistla001 commented:
@petebacondarwin - when can we expect 10.1.0-next.0 to be available?

petebacondarwin commented:
I believe it should be released tomorrow...

JohannesHoppe commented:
I was expecting that with 10.1 we would also get the extraction of tagged template strings:

```
export class AppComponent {
  title = $localize`:meaning|description@@HELLO_WORLD_TS:Hello World HTML (untranslated)`;
}
```

But `ng xi18n --format=xmb` with ` 10.1.0-next.0` still does not extract such messages at all. 😞 
Can we still count on this for 10.1?

petebacondarwin commented:
@JohannesHoppe : The CLI plans to have this integrated for 10.1.0

JohannesHoppe commented:
@petebacondarwin Many thanks for this information!

diegorodridev commented:
Are there any news on this? I am trying to use the localize-extract for templates in my project and it does not extract any messages either.

I have followed the steps you recommended @petebacondarwin : installing the 10.1.0-next... and referencing the @angular/localize to the github repository.

wenqi73 commented:
Updated to `10.1.0-next.4` and run 
```bash
ng build
localize-extract -o src/locale/zh.json -s 'dist/**/*.js' -f json -l zh
```
and it's extracted successfully.

petebacondarwin commented:
This is what I am suggesting as a workaround for now. Add the following to your package.json:

```
{
  ...
  "scripts": {
    ...
    "preextract": "ngc -p tsconfig.app.json",
    "extract": "localize-extract -s \"out-tsc/**/*.js\" -f xlf -o src/locales/messages.xlf"
    ...
  },
  ...
}
```

This will quickly compile the application TS files using the Angular compiler (including adding in the `$localize` calls) and then extract those calls into the `messages.xlf` file. You might need to tweak the options for your set up, and also I recommend ensuring that the `tsconfig.app.json` is setup to generate source-maps, since this will ensure that you get the correct original context for the translations.

diegorodridev commented:
Thanks to both of you! @petebacondarwin @wenqi73

rsivan commented:
> This is what I am suggesting as a workaround for now. Add the following to your package.json:
> 
> ```
> {
>   ...
>   "scripts": {
>     ...
>     "preextract": "ngc -p tsconfig.app.json",
>     "extract": "localize-extract -s \"out-tsc/**/*.js\" -f xlf -o src/locales/messages.xlf"
>     ...
>   },
>   ...
> }
> ```
> 
> This will quickly compile the application TS files using the Angular compiler (including adding in the `$localize` calls) and then extract those calls into the `messages.xlf` file. You might need to tweak the options for your set up, and also I recommend ensuring that the `tsconfig.app.json` is setup to generate source-maps, since this will ensure that you get the correct original context for the translations.

Thanks. This works for me for HTML templates but not for `$localize` strings in Typescript files. Is that expected to change soon, or is there a different function planned for localization of ts strings?.
Thanks.

petebacondarwin commented:
Really? It should work for `$localize` in application code too.
Can you give an example of it not working?

rsivan commented:
Sorry, my bad. It seems the script for `extract` does not override an existing messages.xlf file, so my test case was corrupted.
I changed it now to `-o src/locales/messages2.xlf"` and it does work as it should!

s4m0r4m4 commented:
There is a pointer to this from the documentation (see the blue box under this heading: https://angular.io/guide/i18n#extract-the-source-language-file), since this is closed I imagine that documentation should get updated too.

petebacondarwin commented:
Strictly we need to wait for the CLI to release a version that uses this new extractor rather than the old ViewEngine one.
The integration has been merged to master but is hidden behind an `--ivy` flag: https://github.com/angular/angular-cli/commit/a5293fe74784d84374a2eca47818c7c5c83a7209

Please do go and give it a try when the release candidate for 10.1.0 is published - probably tomorrow.

vekunz commented:
It seems that this is already implemented since 10.1,0-next.6 https://github.com/angular/angular-cli/releases/tag/v10.1.0-next.6

rsivan commented:
I can confirm these issues are solved in @angular/cli and core: 10.1.0-rc.0 when using the `--ivy` flag:

- Bug #37776 which I reported.

- TS $localize`` statements generate expected translation entries. However the context is incorrect, as it refers to the associated HTML file:
```
      <trans-unit id="493297021352091070" datatype="html">
        <source>hello friend</source>
        <context-group purpose="location">
          <context context-type="sourcefile">src/app/view2/view2.component.html</context>
          <context context-type="linenumber">1</context>
        </context-group>
      </trans-unit>
```

sistla001 commented:
I see 10.1.0 is released, so upgraded my project and did `ng xi18n project --format xlf --output-path src/locale`, this command didn't translate translations from typescript, but including --ivy extracted translations from typescript. Is this expected? I have other translations from Kendo libraries, these are getting extracted without --ivy flag but not with --ivy flag. I have "postinstall": "ngcc" and didn't help. Any thoughts?

jaska45 commented:
Extraction from .ts seems working when using --ivy. However the extracted XLIFF no longer contain category="location" unless either meaning, comment or both as passed ti i18n.

petebacondarwin commented:
At the moment we are hiding the new extractor behind the `--ivy` flag until we have ironed out all the bugs.

@jaska45 - please can you create a new issue with a reproduction of your problem?

@sistla001 - currently the new message extractor only looks at the application code - not library code. We need to come up with a proper story for this. In the meantime you could extract them "manually" by running the ivy message extractor directly on the code of the library... Something like:

```
node_modules/.bin/localize-extract -r node_modules/@progress -s */dist/fesm2015/index.js -f xlf -o kendo-messages.xlf
```

sistla001 commented:
Here is what I did to get it to work. I am using xlf-merge to merge kendo-messages.xlf and messages.xlf (project with --ivy flag) into new messages.xlf then using "@ngx-i18nsupport library to generate messages.es.xlf and messages.fr.xlf from messages.xlf.  Does @angular/cli already has support to these functionalities? If not, it would be nice if these too were included. 

```
"extract-i18n": "npm run extract-project && npm run extract-kendo && npm run merge-messages && npm run translate-messages && npm run translate-kendo",
"extract-project": "ng xi18n project --ivy --format xlf --output-path src/locale/xlfs/messages.xlf",
"extract-kendo": "node_modules/.bin/localize-extract  -s node_modules/@progress/**/dist/fesm2015/index.js -f xlf -o src/locale/xlfs/kendo-messages.xlf",
"merge-messages": "xlf-merge src/locale/xlfs/*.xlf -o src/locale/messages.xlf",
"translate-messages": "ng run project:xliffmerge",
"translate-kendo": "npx kendo-translate src/locale/messages.es.xlf --locale es-PR --force && npx kendo-translate src/locale/messages.fr.xlf --locale fr-CA --force",
```

petebacondarwin commented:
That's really cool @sistla001 I had not been aware of the `kendo-translate` tool.

If Kendo provided these messages in a form that we could use (like XLIFF) then you could short-circuit the extraction process and just merge their translations when building the app: you are able to provide an array of translation files for each locale in the `angular.json` file. But it seems that the `kendo-angular-messages` package only contains YAML files that only reference the message's "meaning" rather than its "id".

bastienmoulia commented:
> I am using xlf-merge to merge kendo-messages.xlf and messages.xlf

Maybe you don't have to merge the files because the i18n/locales attributes in angular.json accept an array of filepath.

sistla001 commented:
> > I am using xlf-merge to merge kendo-messages.xlf and messages.xlf
> 
> Maybe you don't have to merge the files because the i18n/locales attributes in angular.json accept an array of filepath.
> 
> ```
> "i18n": {
>   "locales": {
>     "fr": ["src/locale/messages.fr.xlf", "src/locale/kendo-messages.fr.xlf"]
>   }
> }
> ```

Good to know.. Thanks!!

manzonif commented:
Is it possible that the meaning and description in the templates are not extracted? The source file also does not appear.
The $localize instead is extracted as expected.

`"extract-i18n2": "ng xi18n project --ivy --format xlf2 --output-path src/i18n",`

```html
<button i18n="Follow User button|click to follow user@@FollowUser">Follow User</button>
``` 
give me this result:
```xml
    <unit id="FollowCompany">
      <segment>
        <source>Follow Company</source>
      </segment>
    </unit>
```
```typescript
 const title = $localize`:welcome message|home page title@@HomeTitle:You are Welcome`;
```
this:

```xml
    <unit id="HomeTitle">
      <notes>
        <note category="location">src/app/main/pages/home/home.component.js:65</note>
        <note category="description">home page title</note>
        <note category="meaning">welcome message</note>
      </notes>
      <segment>
        <source>Welcome to Expometals</source>
      </segment>
    </unit>
```

petebacondarwin commented:
I just tried that template HTML with the latest Angular and CLI releases (i.e. 10.1.1) and I got:

```xml
    <unit id="FollowUser">
      <notes>
        <note category="location">src/app/app.component.html:3</note>
        <note category="description">click to follow user</note>
        <note category="meaning">Follow User button</note>
      </notes>
      <segment>
        <source>Follow User</source>
      </segment>
    </unit>
```

Can you check your versions @manzonif

manzonif commented:
@petebacondarwin thanks, now it work correctly.

sistla001 commented:
Is there a way to translate text from [title] from HTML without moving the condition to typescript? There is no reason why I shouldn't move the condition to typescript but just curious to know if there is a way.

`<custom-component [title]="condition?'Home':'Dashboard'"></custom-component>`

I tried - i18n-title but didn't extract any messages.

petebacondarwin commented:
@sistla001 - We do not support i18n messages on dynamic input bindings.

jaska45 commented:
@sistla001You have to handle the string in .ts code. For example in your component code

```
title = condition ? $localize`Home` : $localize`Dashboard`;
```
Then you use the title variable in your template

```
<custom-component [title]="{{title}}"></custom-component>
```

JanneHarju commented:
How about normal plural syntax? Is It still working? So in html title="here some plural syntax" i18n.title="@@someTitle"

petebacondarwin commented:
@JanneHarju - what do you mean by "plural syntax"? Do you mean a plural ICU expression? e.g.

```
<div title="{value, plural, =1 {one} other {many}}" i18n.title="@@someTitle">...</div>
```

This still works.

JanneHarju commented:
@petebacondarwin that is exactly what I ment. And its was answer for @sistla001 when he ask is conditional translation possible. With plural syntax it is possible no need to move code to ts side.

sistla001 commented:
> @petebacondarwin that is exactly what I ment. And its was answer for @sistla001 when he ask is conditional translation possible. With plural syntax it is possible no need to move code to ts side.

Thanks @JanneHarju, @petebacondarwin   - I always overlooked this plurals. Thanks for pointing out. It works!!

JanneHarju commented:
How about documentation? When we get it to angular.io? Thanks.

petebacondarwin commented:
Good question! I need to work with @aikidave and @kapunahelewong on this. We will also need to nail down the public API as well.

PowerKiKi commented:
In my migration from Angular 9 to Angular 10, I am trying leverage `$localize` to translate string from components.

I noticed that running `ng xi18n` will extract those string with the new ID format (19 digits), instead of the legacy format (40 hexa). Even though templates string keep using the legacy IDs. This is problematic because code and templates may share the same string, so I would like them to share the same ID, to avoid to have to translate them twice.

In order to avoid this situation, I'd like to permanently migrate everything to the new ID format. But I failed to find a way to specify this option to cli tool `ng xi18n`, or in `angular.json`.

I did find the `--useLegacyIds` flag on the lower level tool. So I can run something like:

```sh
./node_modules/.bin/localize-extract --locale fr --useLegacyIds false --source dist/my-lib/fesm2015/my-lib.js --format xlf --outputPath dist/messages.xlf
```

Is there a more proper way to opt-in the new ID format, without tinkering with low-level things ?

petebacondarwin commented:
Actually there is! Currently the CLI `xi18n` job does not support the `--useLegacyIds` option.

But If you disable legacy message ids in the Angular compiler it will never generate them in the first place, and so the message extractor will have to use the non-legacy ids.

**tsconfig.json**
```jsonc
{
  ...
  "angularCompilerOptions": {
    "enableI18nLegacyMessageIdFormat": false
  }
}
```

Check out this repo: https://github.com/petebacondarwin/issue32912-i18n-no-legacy-test

Note that ngcc always enables legacy message ids at the moment, so any localized messages in templates of library components will still have the old message ids.

PowerKiKi commented:
Thank you for your quick reply. That allowed me to keep using the standard `ng xi18n`, while already moving to latest ID format.

Our library will now ship their own translations as part of their package, and the consuming application will includes the application translation and the library translation, via the array configuration you mentioned in https://github.com/angular/angular/pull/32912#issuecomment-687217233. So, for us, the ngcc part is irrelevant, but it's always good to know.

sistla001 commented:
> @JanneHarju - what do you mean by "plural syntax"? Do you mean a plural ICU expression? e.g.
> 
> ```
> <div title="{value, plural, =1 {one} other {many}}" i18n.title="@@someTitle">...</div>
> ```
> 
> This still works.

@petebacondarwin, @JanneHarju - So I tried the ICU or the plural syntax. 
this works:
<div i18n>{value, plural, =1 {one} other {many}}</div>

but this doesn't work: It displays the full plural syntax
<input value="{value, plural, =1 {one} other {many}}" i18n-value="@@someTitle"/>

example:
https://stackblitz.com/edit/angular-plurals-angular9-5xkaza?file=src%2Fapp%2Fapp.component.html

petebacondarwin commented:
@sistla001 - this is a known issue: https://github.com/angular/angular/issues/21615

sistla001 commented:
Is there any angular recommended approach to support multi-language support for Angular Libraries? I am thinking of creating messages files for each library and them merge these message files in the SPA.

PowerKiKi commented:
This is **not** an official recommendation, but it seems to be working for us. Our lib maintain [localized messages](https://github.com/Ecodev/natural/tree/master/projects/natural/i18n) (translated via Weblate) and ship them [in the package](https://unpkg.com/browse/@ecodev/natural@23.0.0/i18n/). Then our app consume them with the implicit merging mechanism. Here is an extract of `angular.json` that allow that:

```json
{
    "projects": {
        "my-app": {
            "i18n": {
                "sourceLocale": "fr",
                "locales": {
                    "de": [
                        "node_modules/@ecodev/natural/i18n/messages.de.xlf", 
                        "client/i18n/messages.de.xlf"
                    ],
                    "it": [
                        "node_modules/@ecodev/natural/i18n/messages.it.xlf", 
                        "client/i18n/messages.it.xlf"
                    ],
                    "en": [
                        "node_modules/@ecodev/natural/i18n/messages.en.xlf", 
                        "client/i18n/messages.en.xlf"
                    ]
                }
            }
        }
    }
}
```

petebacondarwin commented:
We are looking into ways that these pre-packaged translations in a library can be automatically discovered by the CLI when translating. It would require some specification in the Angular Package Format, I believe.
